File: packages/cli/src/compose.js
/** Compose migration analyser. Mirrors packages/py/src/docker2wslc/compose.py. */

import { parse as parseYaml } from 'yaml';
import { RULES } from './translate.js';

const KEYS = RULES.compose.keys;

function asList(value) {
  if (value === null || value === undefined) return [];
  if (Array.isArray(value)) return value.map(String);
  if (typeof value === 'object') return Object.entries(value).map(([k, v]) => `${k}=${v}`);
  return [String(value)];
}

// wslc rejects lowercase size units: `512m` -> "Invalid memory argument value".
function upperSize(v) {
  return String(v).replace(/^(\d+(?:\.\d+)?)\s*([kmgt])(b?)$/i,
    (_, n, u, b) => n + u.toUpperCase() + b.toUpperCase());
}

// --stop-timeout takes plain seconds; stop_grace_period is a Go duration.
// Compose allows compound forms like `1m30s`, so a single-unit regex is not
// enough -- `1m30s` must become 90, not pass through unchanged.
const UNIT_SECONDS = { h: 3600, m: 60, s: 1, ms: 0.001, us: 0.000001 };
function stopSeconds(v) {
  const raw = String(v).trim();
  if (/^\d+$/.test(raw)) return raw;
  const parts = [...raw.matchAll(/(\d+(?:\.\d+)?)(ms|us|h|m|s)/g)];
  if (!parts.length || parts.map((p) => p[1] + p[2]).join('') !== raw) return raw;
  const total = parts.reduce((acc, p) => acc + parseFloat(p[1]) * UNIT_SECONDS[p[2]], 0);
  return String(Math.round(total));
}

// Values are joined with spaces into one shell line, so anything containing
// whitespace or shell metacharacters must be quoted or the reader gets a command
// that silently splits into the wrong arguments.
function shellQuote(value) {
  const s = String(value);
  if (s === '') return "''";
  if (!/[\s"'$`\\|&;<>()*?!#~\[\]{}]/.test(s)) return s;
  return `'${s.replace(/'/g, "'\\''")}'`;
}

// Compose test: is a string (shell form) or an array led by CMD / CMD-SHELL / NONE.
function healthCmd(test) {
  if (!Array.isArray(test)) return String(test);
  const parts = test.map(String);
  const head = (parts[0] || '').toUpperCase();
  if (head === 'CMD-SHELL' || head === 'CMD') return parts.slice(1).join(' ');
  if (head === 'NONE') return '';
  return parts.join(' ');
}

function buildRun(name, svc) {
  const args = ['wslc', 'run', '-d', '--name', svc.container_name || name];

  for (const env of asList(svc.environment)) args.push('-e', shellQuote(env));
  for (const f of asList(svc.env_file)) args.push('--env-file', f);
  for (const p of asList(svc.ports)) args.push('-p', String(p).replace(/"/g, ''));
  for (const v of asList(svc.volumes)) args.push('-v', v);
  for (const n of asList(svc.networks)) args.push('--network', n);
  // cap_add / cap_drop / devices / privileged / security_opt / expose are
  // deliberately NOT emitted: wslc 2.9.4 has no such flags, so emitting them
  // yields "Argument name was not recognized". They surface as findings instead.
  for (const l of asList(svc.labels)) args.push('--label', shellQuote(l));
  for (const t of asList(svc.tmpfs)) args.push('--tmpfs', t);
  for (const d of asList(svc.dns)) args.push('--dns', String(d));
  for (const u of asList(svc.ulimits)) args.push('--ulimit', shellQuote(u));

  if (svc.working_dir) args.push('-w', String(svc.working_dir));
  if (svc.user) args.push('-u', String(svc.user));
  if (svc.hostname) args.push('--hostname', String(svc.hostname));
  if (svc.domainname) args.push('--domainname', String(svc.domainname));
  if (svc.shm_size) args.push('--shm-size', upperSize(svc.shm_size));
  if (svc.mem_limit) args.push('-m', upperSize(svc.mem_limit));
  if (svc.cpus) args.push('--cpus', String(svc.cpus));
  if (svc.stop_signal) args.push('--stop-signal', String(svc.stop_signal));
  if (svc.stop_grace_period) args.push('--stop-timeout', stopSeconds(svc.stop_grace_period));
  // healthcheck: maps flag-for-flag onto wslc run --health-* (verified 2.9.4.0).
  const hc = svc.healthcheck;
  if (hc && typeof hc === 'object') {
    if (hc.disable === true || hc.disable === 'true') {
      args.push('--no-healthcheck');
    } else {
      if (hc.test) args.push('--health-cmd', shellQuote(healthCmd(hc.test)));
      if (hc.interval) args.push('--health-interval', String(hc.interval));
      if (hc.retries) args.push('--health-retries', String(hc.retries));
      if (hc.timeout) args.push('--health-timeout', String(hc.timeout));
      if (hc.start_period) args.push('--health-start-period', String(hc.start_period));
    }
  }
  if (svc.entrypoint) {
    args.push('--entrypoint', shellQuote(Array.isArray(svc.entrypoint) ? svc.entrypoint.join(' ') : String(svc.entrypoint)));
  }
  if (svc.stdin_open) args.push('-i');
  if (svc.tty) args.push('-t');

  let image = svc.image;
  if (!image && svc.build) image = `${name}:local`;
  args.push(String(image || '<image>'));

  if (svc.command) {
    args.push(...(Array.isArray(svc.command) ? svc.command.map(String) : String(svc.command).split(' ')));
  }
  return args.join(' ');
}

export function analyse(text) {
  const report = { parseError: null, prelude: [], services: [] };

  let doc;
  try {
    doc = parseYaml(text) || {};
  } catch (err) {
    report.parseError = `YAML parse error: ${err.message}`;
    report.exitCode = 2;
    return report;
  }

  if (typeof doc !== 'object' || Array.isArray(doc)) {
    report.parseError = 'Compose file did not parse to a mapping.';
    report.exitCode = 2;
    return report;
  }

  for (const vol of Object.keys(doc.volumes || {})) report.prelude.push(`wslc volume create ${vol}`);
  for (const net of Object.keys(doc.networks || {})) report.prelude.push(`wslc network create ${net}`);

  const services = doc.services || {};
  if (typeof services !== 'object' || Array.isArray(services)) {
    report.parseError = '`services` is not a mapping.';
    report.exitCode = 2;
    return report;
  }

  let hasDependsOn = false;
  for (const [name, svc] of Object.entries(services)) {
    if (!svc || typeof svc !== 'object') continue;
    const sr = { name, command: '', ok: [], partial: [], missing: [], unknown: [] };
    for (const key of Object.keys(svc)) {
      const spec = KEYS[key];
      if (!spec) sr.unknown.push(key);
      else if (spec.status === 'ok') sr.ok.push(key);
      else if (spec.status === 'partial') sr.partial.push({ key, note: spec.note || '' });
      else {
        sr.missing.push({ key, note: spec.note || '' });
        if (key === 'depends_on') hasDependsOn = true;
      }
    }
    sr.command = buildRun(name, svc);
    report.services.push(sr);
  }

  if (hasDependsOn) {
    report.prelude.push(`# start order matters: ${report.services.map((s) => s.name).join(' -> ')}`);
  }

  report.exitCode = report.services.some((s) => s.missing.length)
    ? 2
    : report.services.some((s) => s.partial.length)
      ? 1
      : 0;
  return report;
}

File: packages/py/src/docker2wslc/compose.py
"""Compose migration analyser: docker-compose.yml -> wslc run equivalents."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

from .translate import RULES

COMPOSE_KEYS: dict[str, dict[str, Any]] = RULES["compose"]["keys"]


@dataclass
class ServiceReport:
    name: str
    command: str = ""
    ok: list[str] = field(default_factory=list)
    partial: list[tuple[str, str]] = field(default_factory=list)
    missing: list[tuple[str, str]] = field(default_factory=list)
    unknown: list[str] = field(default_factory=list)


@dataclass
class ComposeReport:
    services: list[ServiceReport] = field(default_factory=list)
    prelude: list[str] = field(default_factory=list)
    parse_error: str | None = None

    @property
    def exit_code(self) -> int:
        if self.parse_error:
            return 2
        if any(s.missing for s in self.services):
            return 2
        if any(s.partial for s in self.services):
            return 1
        return 0

    def as_dict(self) -> dict[str, Any]:
        return {
            "parseError": self.parse_error,
            "prelude": self.prelude,
            "services": [
                {
                    "name": s.name,
                    "command": s.command,
                    "ok": s.ok,
                    "partial": [{"key": k, "note": n} for k, n in s.partial],
                    "missing": [{"key": k, "note": n} for k, n in s.missing],
                    "unknown": s.unknown,
                }
                for s in self.services
            ],
            "exitCode": self.exit_code,
        }


def _as_list(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(v) for v in value]
    if isinstance(value, dict):
        return [f"{k}={v}" for k, v in value.items()]
    return [str(value)]


def _upper_size(value: Any) -> str:
    """wslc rejects lowercase size units: 512m -> Invalid memory argument value."""
    return re.sub(
        r"^(\d+(?:\.\d+)?)\s*([kmgtKMGT])(b?|B?)$",
        lambda m: m.group(1) + m.group(2).upper() + m.group(3).upper(),
        str(value),
    )


_UNIT_SECONDS = {"h": 3600, "m": 60, "s": 1, "ms": 0.001, "us": 0.000001}


def _stop_seconds(value: Any) -> str:
    """--stop-timeout takes plain seconds; stop_grace_period is a Go duration.

    Compose allows compound forms like `1m30s`, so a single-unit regex is not
    enough -- `1m30s` must become 90, not pass through unchanged.
    """
    raw = str(value).strip()
    if re.fullmatch(r"\d+", raw):
        return raw
    parts = re.findall(r"(\d+(?:\.\d+)?)(ms|us|h|m|s)", raw)
    if not parts or "".join(n + u for n, u in parts) != raw:
        return raw
    total = sum(float(n) * _UNIT_SECONDS[u] for n, u in parts)
    return str(round(total))


def _shell_quote(value: Any) -> str:
    """Args are joined into one shell line, so whitespace must be quoted."""
    s = str(value)
    if s == "":
        return "''"
    if not re.search(r"""[\s"'$`\\|&;<>()*?!#~\[\]{}]""", s):
        return s
    return "'" + s.replace("'", "'\\''") + "'"


def _health_cmd(test: Any) -> str:
    """Compose test: is a string, or a list led by CMD / CMD-SHELL / NONE."""
    if not isinstance(test, list):
        return str(test)
    parts = [str(p) for p in test]
    head = (parts[0] if parts else "").upper()
    if head in ("CMD-SHELL", "CMD"):
        return " ".join(parts[1:])
    if head == "NONE":
        return ""
    return " ".join(parts)


def _build_run(name: str, svc: dict[str, Any]) -> str:
    args: list[str] = ["wslc", "run", "-d", "--name", svc.get("container_name") or name]

    for env in _as_list(svc.get("environment")):
        args += ["-e", _shell_quote(env)]
    for env_file in _as_list(svc.get("env_file")):
        args += ["--env-file", env_file]
    for port in _as_list(svc.get("ports")):
        args += ["-p", port.strip('"')]
    for vol in _as_list(svc.get("volumes")):
        args += ["-v", vol]
    for net in _as_list(svc.get("networks")):
        args += ["--network", net]
    # cap_add / cap_drop / devices / privileged / security_opt / expose are
    # deliberately NOT emitted: wslc 2.9.4 has no such flags, so emitting them
    # yields "Argument name was not recognized". They surface as findings.
    for label in _as_list(svc.get("labels")):
        args += ["--label", _shell_quote(label)]
    for tmp in _as_list(svc.get("tmpfs")):
        args += ["--tmpfs", tmp]
    for dns in _as_list(svc.get("dns")):
        args += ["--dns", str(dns)]
    for ulimit in _as_list(svc.get("ulimits")):
        args += ["--ulimit", _shell_quote(ulimit)]

    if svc.get("working_dir"):
        args += ["-w", str(svc["working_dir"])]
    if svc.get("user"):
        args += ["-u", str(svc["user"])]
    if svc.get("hostname"):
        args += ["--hostname", str(svc["hostname"])]
    if svc.get("domainname"):
        args += ["--domainname", str(svc["domainname"])]
    if svc.get("shm_size"):
        args += ["--shm-size", _upper_size(svc["shm_size"])]
    if svc.get("mem_limit"):
        args += ["-m", _upper_size(svc["mem_limit"])]
    if svc.get("cpus"):
        args += ["--cpus", str(svc["cpus"])]
    if svc.get("stop_signal"):
        args += ["--stop-signal", str(svc["stop_signal"])]
    if svc.get("stop_grace_period"):
        args += ["--stop-timeout", _stop_seconds(svc["stop_grace_period"])]
    # healthcheck maps flag-for-flag onto wslc run --health-* (verified 2.9.4.0).
    hc = svc.get("healthcheck")
    if isinstance(hc, dict):
        if hc.get("disable") in (True, "true"):
            args.append("--no-healthcheck")
        else:
            if hc.get("test"):
                args += ["--health-cmd", _shell_quote(_health_cmd(hc["test"]))]
            if hc.get("interval"):
                args += ["--health-interval", str(hc["interval"])]
            if hc.get("retries"):
                args += ["--health-retries", str(hc["retries"])]
            if hc.get("timeout"):
                args += ["--health-timeout", str(hc["timeout"])]
            if hc.get("start_period"):
                args += ["--health-start-period", str(hc["start_period"])]
    if svc.get("entrypoint"):
        ep = svc["entrypoint"]
        args += ["--entrypoint", _shell_quote(ep if isinstance(ep, str) else " ".join(ep))]
    if svc.get("stdin_open"):
        args.append("-i")
    if svc.get("tty"):
        args.append("-t")

    image = svc.get("image")
    if not image and svc.get("build"):
        image = f"{name}:local"
    args.append(str(image or "<image>"))

    cmd = svc.get("command")
    if cmd:
        args += cmd.split() if isinstance(cmd, str) else [str(c) for c in cmd]

    return " ".join(args)


def analyse(text: str) -> ComposeReport:
    """Analyse a compose file. Requires PyYAML; reports cleanly if absent."""
    report = ComposeReport()
    try:
        import yaml
    except ImportError:  # pragma: no cover
        report.parse_error = "PyYAML is required: pip install 'docker2wslc[yaml]'"
        return report

    try:
        doc = yaml.safe_load(text) or {}
    except Exception as exc:
        report.parse_error = f"YAML parse error: {exc}"
        return report

    if not isinstance(doc, dict):
        report.parse_error = "Compose file did not parse to a mapping."
        return report

    for vol in (doc.get("volumes") or {}):
        report.prelude.append(f"wslc volume create {vol}")
    for net in (doc.get("networks") or {}):
        report.prelude.append(f"wslc network create {net}")

    services = doc.get("services") or {}
    if not isinstance(services, dict):
        report.parse_error = "`services` is not a mapping."
        return report

    for name, svc in services.items():
        if not isinstance(svc, dict):
            continue
        sr = ServiceReport(name=str(name))
        for key in svc:
            spec = COMPOSE_KEYS.get(str(key))
            if spec is None:
                sr.unknown.append(str(key))
            elif spec["status"] == "ok":
                sr.ok.append(str(key))
            elif spec["status"] == "partial":
                sr.partial.append((str(key), spec.get("note", "")))
            else:
                sr.missing.append((str(key), spec.get("note", "")))
        sr.command = _build_run(str(name), svc)
        report.services.append(sr)

    # depends_on ordering hint
    ordered = [s.name for s in report.services]
    if any("depends_on" in dict(s.missing) for s in report.services):
        report.prelude.append(f"# start order matters: {' -> '.join(ordered)}")

    return report
